### DISCUSSION TITLE
antd@ 5.x 如何将px转换为 rem,  为什么mui emotion可以？难道阿里团队就没有考虑到

### DISCUSSION THREAD

**Discussion by AsuraLuo** [ant-design/ant-design#40789] [General]
1. Mui @emotion-react createCahe都支持stylisPlugins插件配置，可以px转rem, vw, vh, antd5都不支持
2. 希望antd5能够支持在Turbo Repo Nextjs项目里面的SSR, 目前antd5也不支持
3. antd5, mui都是基于emotion开发的方案，为什么差别这么大，经不住turbo repo, px ==> rem的考验

```javascript
createCache({
    key: 'mui-style',
    prepend: true,
    insertionPoint,
    stylisPlugins: [stylisPx2RemPlugin(options)]
  })
```

**Comment by zombieJ** [comment]
antd 底层用的不是 emotion，所以要实现 px 转 rem 会不太一样。可以做一个 transformer 来实现转译，有兴趣来个 PR 吗？可以参考：https://github.com/ant-design/cssinjs/blob/master/src/transformers/legacyLogicalProperties.ts

**Reply by AsuraLuo** [reply]
@zombieJ , 希望antd5能够支持在Turbo Repo Nextjs项目里面的SSR, 目前antd5也不支持
这个问题能解决吗？
https://github.com/ant-design/cssinjs/issues/86

**Reply by AsuraLuo** [reply]
我看antd源码package.json,   这里面有依赖emotion
"@emotion/babel-preset-css-prop": "^11.10.0",
    "@emotion/css": "^11.10.5",
    "@emotion/react": "^11.10.4",
    "@emotion/server": "^11.4.0",

**Reply by li-jia-nan** [reply]
package.json 里面有 emotion 是因为官网是用 emotion 写的

**Reply by zombieJ** [reply]
> 我看antd源码package.json, 这里面有依赖emotion "@emotion/babel-preset-css-prop": "^11.10.0", "@emotion/css": "^11.10.5", "@emotion/react": "^11.10.4", "@emotion/server": "^11.4.0",

devDeps 里还包含官网的依赖，官网使用 emotion 开发来测试 `@ant-design/cssinjs` 和 `emotion` 的兼容性哈

**Reply by zombieJ** [reply]
> @zombieJ , 希望antd5能够支持在Turbo Repo Nextjs项目里面的SSR, 目前antd5也不支持 这个问题能解决吗？ [ant-design/cssinjs#86](https://github.com/ant-design/cssinjs/issues/86)

@MadCcc  看一下这个~

**Reply by AsuraLuo** [reply]
@MadCcc ，麻烦看下这个，是不是和pnpm软连接有问题
https://github.com/ant-design/cssinjs/issues/86

**Reply by Wxh16144** [reply]
> 有兴趣来个 PR 吗

有什么进展么，我愿意去尝试一下~

**Reply by RedJue** [reply]
> @MadCcc ，麻烦看下这个，是不是和pnpm软连接有问题
> 
> https://github.com/ant-design/cssinjs/issues/86

最好别用pnpm，兼容性问题还不小，推荐使用cnpm

**Reply by AsuraLuo** [reply]
@Wxh16144 , @MadCcc , @zombieJ 可能阿里这边还需要修改源代码，我自己PR了一个transform.ts，但是有些变量的px是在代码里面写死的，tranform拿不到，而且css-in-js的包无法在web spa里面修改，大坑

```javascript
import { CSSObject, Transformer } from '@ant-design/cssinjs'

export const transform: Transformer = {
  visit: (cssObj) => {
    const clone: CSSObject = {}

    Object.keys(cssObj).forEach((key) => {
      const value: any = cssObj[key]
      const remSize: number = 16

      if (typeof value === 'number' || typeof value === 'string') {
        const pxRegexp: RegExp = /"[^"]+"|'[^']+'|url\([^)]+\)|(\d*\.?\d+)px/g
        const expression: string = String(value).replace(
          pxRegexp,
          (match, group) => {
            return group ? `${Number(group) / remSize}rem` : match
          }
        )
        clone[key] = typeof value === 'number' ? value : expression
      } else {
        clone[key] = value
      }
    })

    return clone
  }
}
```

**Reply by AsuraLuo** [reply]
typeof value === 'number' ，当变量是数字，比如32, 无法转换，因为代码写死的px单位。只有变量为'12px'，这种能转化。但是web端 SPA这个tranform.ts都不会执行，这个是为什么?

**Reply by Wxh16144** [reply]
> 我自己PR了一个transform.ts

内容不要做重复了哦，我已经开了一个 [PR#89](https://github.com/ant-design/cssinjs/pull/89).

> typeof value === 'number' ，当变量是数字，比如32, 无法转换

可以在上面的 PR 添加 case， 然后加个逻辑判断一下， 🤔

**Reply by AsuraLuo** [reply]
@Wxh16144 @MadCcc @zombieJ 还有个问题，为什么transform在SPA客户端里面无法执行

**Reply by Wxh16144** [reply]
> 为什么transform在SPA客户端里面无法执行

需要提供一个最小复现 repo，否则我们没办法帮助你排查问题。

**Reply by zombieJ** [reply]
> @Wxh16144 @MadCcc @zombieJ 还有个问题，为什么transform在SPA客户端里面无法执行

需要重现 +1，大概率是包的地方不对导致没有读到 context 导致的。

**Reply by AsuraLuo** [reply]
问题已经解决，包括turbo repo ssr问题, thanks，和antd版本升级到最新问题解决
